pkg/kubelet/volume/nfs/nfs.go

```diff
@@ -0,0 +1,200 @@
+}
+
+func (plugin *nfsPlugin) Name() string {
+	return nfsPluginName
+}
+
+func (plugin *nfsPlugin) CanSupport(spec *api.Volume) bool {
+	if spec.VolumeSource.NFS != nil {
+		return true
+	}
+	return false
+}
+
+func (plugin *nfsPlugin) NewBuilder(spec *api.Volume, podUID types.UID) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, podUID, NewNFSMounter())
+}
+
+func (plugin *nfsPlugin) newBuilderInternal(spec *api.Volume, podUID types.UID, mounter NFSMountInterface) (volume.Builder, error) {
+	return &nfs{
+		volName:      spec.Name,
+		server:       spec.VolumeSource.NFS.Server,
+		exportPath:   spec.VolumeSource.NFS.SourcePath,
+		mountOptions: spec.VolumeSource.NFS.MountOptions,
+		mounter:      mounter,
+		podUID:       podUID,
+		plugin:       plugin,
+	}, nil
+}
+
+func (plugin *nfsPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, NewNFSMounter())
+}
+
+func (plugin *nfsPlugin) newCleanerInternal(volName string, podUID types.UID, mounter NFSMountInterface) (volume.Cleaner, error) {
+	return &nfs{
+		volName:      volName,
+		server:       "",
+		exportPath:   "",
+		mountOptions: "",
+		mounter:      mounter,
+		podUID:       podUID,
+		plugin:       plugin,
+	}, nil
+}
+
+// NFS volumes represent a bare host file or directory mount of an NFS export.
+type nfs struct {
+	volName      string
+	podUID       types.UID
+	server       string
+	exportPath   string
+	mountOptions string
+	mounter      NFSMountInterface
+	plugin       *nfsPlugin
+}
+
+func (nfs *nfs) SetUp() error {
+	mountpoint, err := mount.IsMountPoint(nfs.GetPath())
+	glog.V(4).Infof("NFS mount set up: %s %v %v", nfs.GetPath(), mountpoint, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if mountpoint {
+		return nil
+	}
+	exportDir := nfs.exportPath
+	mountDir := nfs.GetPath()
+	exists, err := exists(mountDir)
+	if err != nil {
+		glog.Errorf("Mount directory is invalid: %v, %v", mountDir, err)
+	}
+	if !exists {
+		os.MkdirAll(mountDir, 0750)
+	}
+	err = nfs.mounter.Mount(nfs.server, exportDir, mountDir, nfs.mountOptions)
+	if err != nil {
+		mountpoint, mntErr := mount.IsMountPoint(nfs.GetPath())
+		if mntErr != nil {
+			glog.Errorf("IsMountpoint check failed: %v", mntErr)
+			return err
+		}
+		if mountpoint {
+			if mntErr = nfs.mounter.Unmount(nfs.GetPath()); mntErr != nil {
+				glog.Errorf("Failed to unmount: %v", mntErr)
+				return err
+			}
+			mountpoint, mntErr := mount.IsMountPoint(nfs.GetPath())
+			if mntErr != nil {
+				glog.Errorf("IsMountpoint check failed: %v", mntErr)
+				return err
+			}
+			if mountpoint {
+				// This is very odd, we don't expect it.  We'll try again next sync loop.
+				glog.Errorf("%s is still mounted, despite call to unmount().  Will try again next sync loop.", nfs.GetPath())
+				return err
+			}
+		}
+		os.Remove(nfs.GetPath())
+		return err
+	}
+	return nil
+}
+
+func exists(path string) (bool, error) {
```

Comment:
It's not really a mount specific command, so I'm not sure it should go in there. It's also wrapping two os package calls and not really algorithmic in nature, so not sure if it's of general use.